supervisor/src/bin/supervisor-encore.rs

```diff
@@ -0,0 +1,278 @@
+                        .hosted_gateways
+                        .iter()
+                        .map(|rid| {
+                            gateways
+                                .iter()
+                                .find(|g| g.rid == *rid)
+                                .expect("Gateway rid not found in infra resources")
+                                .encore_name
+                                .clone()
+                        })
+                        .collect::<Vec<String>>(),
+                )
+            }
+            Err(_) => {
+                // If protobuf decoding fails, try JSON decoding
+                let config: RuntimeConfig = serde_json::from_slice(&runtime_config)
+                    .expect("Failed to parse RuntimeConfig as JSON");
+                (
+                    config.hosted_services,
+                    config.gateways.iter().map(|g| g.name.clone()).collect(),
+                )
+            }
+        };
+
+    // Get the exposed port from the environment variable, defaulting to 8080 if not set
+    let exposed_port = env::var("PORT")
+        .ok()
+        .and_then(|p| p.parse::<u16>().ok())
+        .unwrap_or(8080);
+
+    // Assign a unique port to each hosted service
+    let mut service_ports = std::collections::HashMap::new();
+
+    // Start assigning services to ports from the exposed port (if there are gateways, start at +1)
+    let mut port: u16 = exposed_port + if hosted_gateways.is_empty() { 0 } else { 1 };
+    for service in &hosted_services {
+        service_ports.insert(service.clone(), port);
+        port += 1;
+    }
+
+    // Run all gateways on the first port
+    let mut procs = Vec::new();
+    if !hosted_gateways.is_empty() {
+        procs.push(
+            create_process_config(
+                vec![],
+                hosted_gateways,
+                exposed_port,
+                &service_ports,
+                &supervisor_config,
+            )
+            .expect("Failed to create process for gateways"),
+        );
+    }
+
+    // Create a process for each service and assign it to the selected port
+    for (service_name, service_port) in &service_ports {
+        procs.push(
+            create_process_config(
+                vec![service_name.clone()],
+                vec![],
+                *service_port,
+                &service_ports,
+                &supervisor_config,
+            )
+            .expect("Failed to create process for service"),
+        );
+    }
+
+    let sv = Supervisor::new(procs);
+    let root_token = CancellationToken::new();
+    let supervisor_token = root_token.child_token();
+
+    // Spawn a task to listen for SIGINT or SIGTERM and cancel the root token
+    tokio::spawn(async move {
+        // Wait for SIGINT or SIGTERM
+        tokio::signal::ctrl_c()
+            .await
+            .expect("Failed to listen for ctrl+c");
+        println!("Received shutdown signal. Initiating graceful shutdown...");
```

Comment:
you could use `log::info!("hello")` here to use the logger instead of just writing to stdout